iac-extensions/arm/src/main/java/org/sonar/iac/arm/checks/IpRestrictedAdminAccessCheck.java

```diff
@@ -24,57 +24,64 @@
-  private static final String PROPERTIES = "properties";
-  private static final String NETWORK_SECURITY_GROUP = "networkSecurityGroup";
-  private static final String SUBNETS = "subnets";
-  private static final String SUBNET = "subnet";
+@Rule(key = "S6321")
+public class IpRestrictedAdminAccessCheck extends AbstractArmResourceCheck {
   private static final String ARRAY_TOKEN = "*";
-  private static final Map<String, List<String>> PATH_PER_RESOURCE_TYPE = Map.of(
-    "Microsoft.Network/networkSecurityGroups/securityRules", List.of(),
-    "Microsoft.Network/networkSecurityGroup", List.of(SECURITY_RULES, ARRAY_TOKEN, PROPERTIES),
-    "Microsoft.Network/virtualNetworks/subnets", List.of(NETWORK_SECURITY_GROUP, PROPERTIES, SECURITY_RULES, ARRAY_TOKEN, PROPERTIES),
-    "Microsoft.Network/virtualNetworks", List.of(SUBNETS, ARRAY_TOKEN, PROPERTIES, NETWORK_SECURITY_GROUP, PROPERTIES, SECURITY_RULES, ARRAY_TOKEN, PROPERTIES),
-    "Microsoft.Network/networkInterfaces",
-    List.of("ipConfigurations", ARRAY_TOKEN, PROPERTIES, SUBNET, PROPERTIES, NETWORK_SECURITY_GROUP, PROPERTIES, SECURITY_RULES, ARRAY_TOKEN, PROPERTIES));
   private static final Set<String> SOURCE_ADDRESS_PREFIX_SENSITIVE = Set.of("*", ALL_IPV4, ALL_IPV6, "Internet");
   private static final Set<String> SENSITIVE_PROTOCOL = Set.of("*", "TCP");
 
   @Override
-  public void initialize(InitContext init) {
-    init.register(ResourceDeclaration.class, (ctx, resource) -> {
-      List<String> path = PATH_PER_RESOURCE_TYPE.get(resource.type().value());
-      if (path != null) {
-        List<Tree> listProperties = resolveProperties(new LinkedList<>(path), resource);
-        for (Tree properties : listProperties) {
-          ResourceWithIpRestrictedAdminAccessChecker checker = new ResourceWithIpRestrictedAdminAccessChecker(resource, properties);
-          if (checker.isSensitive()) {
-            checker.reportIssue(ctx);
-          }
-        }
+  protected void registerResourceConsumer() {
+    register("Microsoft.Network/networkSecurityGroups/securityRules",
+      (ctx, resource) -> checkResourcePath(ctx, resource, Collections.emptyList()));
+
+    register("Microsoft.Network/networkSecurityGroup",
+      (ctx, resource) -> checkResourcePath(ctx, resource, propertyPath("securityRules/*/properties")));
+
+    register("Microsoft.Network/virtualNetworks/subnets",
+      (ctx, resource) -> checkResourcePath(ctx, resource, propertyPath("networkSecurityGroup/properties/securityRules/*/properties")));
+
+    register("Microsoft.Network/virtualNetworks",
+      (ctx, resource) -> checkResourcePath(ctx, resource, propertyPath("subnets/*/properties/networkSecurityGroup/properties/securityRules/*/properties")));
+
+    register("Microsoft.Network/networkInterfaces",
+      (ctx, resource) -> checkResourcePath(ctx, resource,
+        propertyPath("ipConfigurations/*/properties/subnet/properties/networkSecurityGroup/properties/securityRules/*/properties")));
+  }
```

Comment:
The change should make the code more readable. This is not super performant since the list will be generated on every resource match. However, we do not expect more than 10 of these resources per analysis, so the consumer may be called less than a dozen times.

Reply:
It's definitely more readable!
I was indeed at first worried a bit about the performance, but as you said there won't be that much resource per analysis, it's different from a typical programming language where we could potentially expect a lot of usage of specific instruction/keywords.